### ISSUE TITLE
[Angular 9.0.4] Building throws `ERROR in Error when flattening the source-map "[...].d.ts.map" for "[...].d.ts": Maximum call stack size exceeded`

### ISSUE THREAD

**Issue by nunocastromartins** [angular/angular#35757]
# 🐞 bug report

### Affected Package

Probably `@angular/cli`, but [I was asked to open the issue here](https://github.com/angular/angular-cli/issues/17106#issuecomment-592477628).

### Is this a regression?

Yes. The issue does not occur with Angular 9.0.2. Version 9.0.3 causes a different `RangeError: Maximum call stack size exceeded` error, tracked here: https://github.com/angular/angular-cli/issues/17106.

### Description

Building and serving both fail consistently with a bunch of the following messages:
```
ERROR in Error when flattening the source-map "[/path/to/file1].d.ts.map" for "[/path/to/file1].d.ts": RangeError: Maximum call stack size exceeded
ERROR in Error when flattening the source-map "[/path/to/file2].d.ts.map" for "[/path/to/file2].d.ts": RangeError: Maximum call stack size exceeded
...
ERROR in Error when flattening the source-map "[/path/to/fileN].d.ts.map" for "[/path/to/fileN].d.ts": RangeError: Maximum call stack size exceeded
```
The issue seems similar to the above mentioned one and deleting `node_modules` does not fix it.

My setup is an Angular workspace with two projects: an app and a lib. The source-maps being "flattened" when the error occurs belong to the lib. The error occurs when attempting to build (`--prod` or otherwise) or serve the app.

## 🌍  Your Environment

**Angular Version:**
```
Angular CLI: 9.0.4
Node: 13.9.0
OS: linux x64

Angular: 9.0.4
... animations, cli, common, compiler, compiler-cli, core, forms
... platform-browser, platform-browser-dynamic, router
Ivy Workspace: Yes

Package                            Version
------------------------------------------------------------
@angular-devkit/architect          0.900.4
@angular-devkit/build-angular      0.900.4
@angular-devkit/build-ng-packagr   0.900.4
@angular-devkit/build-optimizer    0.900.4
@angular-devkit/build-webpack      0.900.4
@angular-devkit/core               9.0.4
@angular-devkit/schematics         9.0.4
@angular/cdk                       9.1.0
@ngtools/webpack                   9.0.4
@schematics/angular                9.0.4
@schematics/update                 0.900.4
ng-packagr                         9.0.2
rxjs                               6.5.4
typescript                         3.7.5
webpack                            4.41.2
```

**Comment by petebacondarwin** [comment]
Thanks @nunocastromartins - I was hoping you could give me some of the files that have caused this error, so that I can investigate.

**Comment by nunocastromartins** [comment]
I'm afraid I can't do that right now, but I will try to set up a minimal reproduction if no one beats me to it. :)

**Comment by otodockal** [comment]
@nunocastromartins @petebacondarwin removing node_modules (...and maybe package-lock.json) helped me. 
reproduction: Try to migrate to 9.0.3 (I could compile) and to 9.0.4.
and I've got the same error...

```bash
> ngcc --properties es2015 browser module main --first-only --create-ivy-entry-points

Compiling @angular/cdk/keycodes : es2015 as esm2015
Error: Error when flattening the source-map "/../node_modules/@angular/cdk/keycodes/index.d.ts.map" for "/../node_modules/@angular/cdk/keycodes/index.d.ts": RangeError: Maximum call stack size exceeded
Compiling @angular/animations : es2015 as esm2015
Compiling @angular/compiler/testing : es2015 as esm2015
Compiling @angular/core : es2015 as esm2015
Compiling @angular/common : es2015 as esm2015
Compiling @angular/cdk/observers : es2015 as esm2015
Compiling @angular/animations/browser : es2015 as esm2015
Error: Error when flattening the source-map "/../node_modules/@angular/cdk/observers/observe-content.d.ts.map" for "/../node_modules/@angular/cdk/observers/observe-content.d.ts": RangeError: Maximum call stack size exceeded
Error: Error when flattening the source-map "/../node_modules/@angular/cdk/observers/index.d.ts.map" for "/../node_modules/@angular/cdk/observers/index.d.ts": RangeError: Maximum call stack size exceeded
Compiling @angular/cdk/platform : es2015 as esm2015
Error: Error when flattening the source-map "/../node_modules/@angular/cdk/platform/platform.d.ts.map" for "/../node_modules/@angular/cdk/platform/platform.d.ts": RangeError: Maximum call stack size exceeded
Error: Error when flattening the source-map "/../node_modules/@angular/cdk/platform/platform-module.d.ts.map" for "/../node_modules/@angular/cdk/platform/platform-module.d.ts": RangeError: Maximum call stack size exceeded
Error: Error when flattening the source-map "/../node_modules/@angular/cdk/platform/index.d.ts.map" for "/../node_modules/@angular/cdk/platform/index.d.ts": RangeError: Maximum call stack size exceeded

```

**Comment by petebacondarwin** [comment]
Great thanks @otodockal - Just to be clear: are you saying that refreshing node_modules fixed this problem or that you still see this error after removing and reinstalling node_modules?

**Comment by Achilles1515** [comment]
I was getting the same error with my project. Deleting then reinstalling node_modules fixed the issue.

**Comment by otodockal** [comment]
@petebacondarwin when I removed **node_modules** and run "npm install" again, it fixed that problem, at least for me. The error is not there anymore.

I removed also package-lock.json, but I don't know if it's relevant...If I should advise to do it.

**Comment by nunocastromartins** [comment]
I could reproduce what I reported consistently even after removing `node_modules`. However, after removing my `dist` folder (generated when building the lib), I can no longer reproduce the issue, even across rebuilds.

This is odd because I believe that Angular shouldn't even be looking at that folder, but that might be a different issue.

I'll close this for now since I can no longer reproduce the issue, but feel free to reopen it if others can reproduce it more consistently. I'll also open a new issue describing what I mean by "Angular shouldn't be looking at that folder" (which I've now realised is causing an annoying error).

**Comment by stephanecot** [comment]
I don't understand, I don't have the error anymore on my local computer (windows), but on my gitlab ci job (docker and linux), I am getting this:

ERROR in Error when flattening the source-map "/builds/applications/frontsxxxx/node_modules/ngx-window-token/ngx-window-token.d.ts.map" for "/builds/applications/fronts/xxxx/node_modules/ngx-window-token/ngx-window-token.d.ts": RangeError: Maximum call stack size exceeded

**Comment by z6tanker** [comment]
I had the same problem. I removed my `dist` and `package-lock.json` and then ran `npm install`.

**Comment by ayyash** [comment]
I don't understand, when you say "dist" folder, what is that? after failure to build that destination folder is not supposed to be there to delete, so what do you all mean by dist folder?

**Comment by nunocastromartins** [comment]
The `dist` folder that was there due to the previous successful build. Somehow it was preventing new builds from succeeding.

**Comment by stephanecot** [comment]
with angular 9, there is a incremental build process which is in the /dist folder.

**Comment by mattiaskagstrom** [comment]
This fails on azure devops when running "npm run build" (ng build) which should remove the output before building since the default on "deleteOutputPath" is true. If we have to remove the package-lock that would defeat the purpose of it..

**Comment by adamdport** [comment]
adamdport's guide to fixing angular gets a brand new entry today!

- Delete node modules
- Delete lock file
- clear npm cache
- `npm rebuild` whatever library is causing grief
- delete AppData cache folder
- **_delete output (dist) folder_**

**Comment by ayyash** [comment]
@adamdport 😊 personally none of that worked for me, what fixed it was updating terser library to latest, but it doesn't have to work for you, so my guide is: grab the code and slam it against the wall a couple of times, what sticks, fix it :p

**Comment by progmancod** [comment]
Same error here, none of that worked for me.

Error: Error when flattening the source-map ".../node_modules/angular-draggable-droppable/bundles/angular-draggable-droppable.umd.js.map" for ".../node_modules/angular-draggable-droppable/bundles/angular-draggable-droppable.umd.js": RangeError: Maximum call stack size exceeded

**Comment by jcompagner** [comment]
i moved from 9.0.2 to 9.0.5 and i had the same problem, but only removing the complete node_modules and doing an npm install again did fix it it seems. (i didnt have a dist folder because of the failed build before that anyway)

I wonder what that then is. what is suddenly changed in node_modules that it works now?

i see here that some are saying that we can have incremental build really also over multiply runs of "npm run build" ? (without bazel?). How would that work?

**Comment by stephanecot** [comment]
> i see here that some are saying that we can have incremental build really also over multiply runs of "npm run build" ? (without bazel?). How would that work?

It's only for "external libs", not your own project and code.

It's when you see:
Compiling @angular/core : es2015 as esm2015
Compiling @angular/common : es2015 as esm2015
Compiling @angular/platform-browser : es2015 as esm2015
Compiling @angular/platform-browser-dynamic : es2015 as esm2015
....

**Comment by DaleyKD** [comment]
My build just failed in Azure DevOps, just as @mattiaskagstrom stated.  Going to have to rollback this changeset.  It built fine locally.

We're using docker, both locally and in the cloud.

**Comment by geoyogesh** [comment]
i resolved this issue by updating node to 10.13

**Comment by mduft** [comment]
Having the same issue updating today. Removing the dist folder fixed the problem. Now I "just" have to tell everyone who will get the update commit :|

**Comment by fischermatte** [comment]
we still experiencing this issue on a fresh checkout with bitbucket pipelines. there are no options to delete any caches etc. is there any explanation where this is coming from? The lib we are importing is  a angular v9 lib which was compiled with "enableIvy" set to false. we are using v9.0.5 with node 12.16.1.

**Comment by stephanecot** [comment]
> we still experiencing this issue on a fresh checkout with bitbucket pipelines. there are no options to delete any caches etc. is there any explanation where this is coming from? The lib we are importing is a angular v9 lib which was compiled with "enableIvy" set to false. we are using v9.0.5 with node 12.16.1.

It's not an Ivy problem. It's the process to "optimize" external libs (see my previous post).
It's when compiling the vendor-xx.js file in the /dist folder.

In gitlab ci, my /dist folder was cached by gitlab runner job.

**Comment by fischermatte** [comment]
> In gitlab ci, my /dist folder was cached by gitlab runner job.

hm, do you mean the folder "/__ivy_ngcc__" ? in my /dist folder is only the outcome of the compiled app itself. in "__ivy_ngcc__" i see some libs.

**Comment by petebacondarwin** [comment]
@fischermatte - according to these docs it is possible to disable a cache in gitlab - https://docs.gitlab.com/ee/ci/caching/#disabling-cache-on-specific-jobs. It might be worth trying this to demonstrate that it is actually something unwanted being stored in the cache between builds. For example, if ngcc is modifying the `node_modules` folder before it was stored in the cache then it could be corrupted and that might be why the new version of ngcc is unable to process it correctly.

Interestingly, I would have expected your cache to be keyed off something like a hash of your `yarn.lock`/`package-json.lock` file so that it was not reused when that changes, to prevent this sort of thing. Interested to know how that is setup.

**Comment by fischermatte** [comment]
> @fischermatte - according to these docs it is possible to disable a cache in gitlab
@petebacondarwin we use bitbucket pipeline without explicit cache being defined in our pipeline config. But - when using node `node_modules` seems to be implicit predefined. Unfortunately bitbucket does not provide a way to invalidate this cache. 

Anyway - even locally on my machine. When I delete everything (`node_modules`, `dist`, `package-lock.json`, `__ivy_ngcc__`) and do fresh install and build I get the error with angular 9.0.5. but not with 9.0.2.

**Comment by msegmx** [comment]
Updated angular (@angular/cdk ../cli ../core) and @ngrx/store and got the same error after running "ng s".

I *only* had to delete the dist folder to make "ng s" work again. 

(If it hadn't worked I would've gone for the node_modules folder, but luckily there was no need for that)

HTH someone.

**Comment by DownGoat** [comment]
Had this problem in Azure Devops as well. Tried doing ``npm ci`` instead of ```npm install```, but that failed as well. Managed to fix this by adding task before ```npm install``` that just deletes ```node_modules```.

**Comment by soysal** [comment]
In my case, it happens only once.  When I retry building again, it works without any problem, even if I haven't done anything.
That's why, some of these solutions may be misleading. 🤔

**Comment by sysmat** [comment]
The prolem also with Angular CLI: 9.1.0 and angular 9

**Comment by petebacondarwin** [comment]
@sysmat - when you say Angular 9, do you mean 9.1.0? The improvements to source-map flattening are in the framework not the CLI

**Comment by sysmat** [comment]
@petebacondarwin yes, sorry

**Comment by petebacondarwin** [comment]
@sysmat - please open a new issue with steps to reproduce the error.

**Comment by francovp** [comment]
I don't have this problem on `9.0.7` but I have it when I update to `9.1.0` :/
